src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1179,6 +1286,210 @@ static void uct_ib_mlx5_devx_md_cleanup(uct_ib_md_t *ibmd)
+
+static ucs_status_t
+uct_ib_mlx5_devx_reg_exported_key(uct_ib_md_t *ib_md, uint64_t flags,
+                                  void *address, size_t length,
+                                  uct_ib_mem_t *ib_memh)
+{
+    uct_ib_mlx5_md_t *md                                              =
+            ucs_derived_of(ib_md, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_mem_t *memh                                           =
+            ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(create_mkey_in)]                = {0};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(create_mkey_out)]              = {0};
+    struct mlx5dv_pd dvpd                                             = {0};
+    struct mlx5dv_obj dv                                              = {{0}};
+    char ein[UCT_IB_MLX5DV_ST_SZ_BYTES(allow_other_vhca_access_in)]   = {0};
+    char eout[UCT_IB_MLX5DV_ST_SZ_BYTES(allow_other_vhca_access_out)] = {0};
+    struct mlx5dv_devx_umem_in umem_in;
+    ucs_status_t status;
+    void *access_key;
+    void *mkc;
+    int rc;
+
+    /* register umem */
+    umem_in.addr        = address;
+    umem_in.size        = length;
+    umem_in.access      = UCT_IB_UMEM_ACCESS;
+    umem_in.pgsz_bitmap = UCS_BIT(12);
+    umem_in.comp_mask   = 0;
+    memh->umem = mlx5dv_devx_umem_reg_ex(md->super.dev.ibv_context, &umem_in);
+    if (memh->umem == NULL) {
+        uct_md_log_mem_reg_error(flags,
+                                 "mlx5dv_devx_umem_reg_ex() failed: %m");
+        status = UCS_ERR_NO_MEMORY;
+        goto err_out;
+    }
+
+    /* obtain pdn */
+    dv.pd.in = md->super.pd;
+    dv.pd.out = &dvpd;
+    rc = mlx5dv_init_obj(&dv, MLX5DV_OBJ_PD);
+    if (rc) {
+        uct_md_log_mem_reg_error(flags, "mlx5dv_init_obj() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    /* create mkey */
+    mkc = UCT_IB_MLX5DV_ADDR_OF(create_mkey_in, in, memory_key_mkey_entry);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, opcode, UCT_IB_MLX5_CMD_OP_CREATE_MKEY);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, translations_octword_actual_size, 1);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, pg_access, 1);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, mkey_umem_valid, 1);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, mkey_umem_id, memh->umem->umem_id);
+    UCT_IB_MLX5DV_SET64(create_mkey_in, in, mkey_umem_offset, 0);
+    UCT_IB_MLX5DV_SET(mkc, mkc, access_mode_1_0, UCT_IB_MLX5_MKC_ACCESS_MODE_MTT);
+    UCT_IB_MLX5DV_SET(mkc, mkc, a, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, rw, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, rr, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, lw, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, lr, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, crossing_target_mkey, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, log_entity_size, 12);
+    UCT_IB_MLX5DV_SET(mkc, mkc, qpn, 0xffffff);
+    UCT_IB_MLX5DV_SET(mkc, mkc, pd, dvpd.pdn);
+    UCT_IB_MLX5DV_SET(mkc, mkc, mkey_7_0, UCT_IB_CROSS_KEY_IDX);
```

Comment:
what is this field?

Reply:
this is the name of the field according to PRM.
and it is `Variant part of MKey specified by this MKey context`.
so, we set CROSS_KEY_IDX here